packages/firestore/test/unit/specs/bundle_spec.test.ts

```diff
@@ -0,0 +1,349 @@
+          .expectEvents(query, {
+            modified: [
+              doc(
+                'collection/a',
+                250,
+                { key: 'patched' },
+                { hasLocalMutations: true }
+              )
+            ],
+            hasPendingWrites: true
+          })
+          .writeAcks('collection/a', 1000)
+          // loading bundleBeforeMutationAck will not raise snapshots, because it is before
+          // the acknowledged mutation.
+          .loadBundle(bundleBeforeMutationAck)
+          // loading bundleAfterMutationAck will raise a snapshot, because it is after
+          // the acknowledged mutation.
+          .loadBundle(bundleAfterMutationAck)
+          .expectEvents(query, {
+            modified: [doc('collection/a', 1001, { key: 'fromBundle' })],
+            fromCache: true
+          })
+      );
+    }
+  );
+
+  specTest(
+    'Newer docs from bundles should keep not raise snapshot if there are unacknowledged writes',
+    [],
+    () => {
+      const query = Query.atPath(path('collection'));
+      const docA = doc('collection/a', 250, { key: 'a' });
+
+      const bundleString = bundleWithDocument({
+        key: docA.key,
+        readTime: 1001,
+        createTime: 250,
+        updateTime: 1001,
+        content: { key: { stringValue: 'fromBundle' } }
+      });
+
+      return (
+        spec()
+          .withGCEnabled(false)
+          .userListens(query)
+          .watchAcksFull(query, 250, docA)
+          .expectEvents(query, {
+            added: [doc('collection/a', 250, { key: 'a' })]
+          })
+          .userPatches('collection/a', { key: 'patched' })
+          .expectEvents(query, {
+            modified: [
+              doc(
+                'collection/a',
+                250,
+                { key: 'patched' },
+                { hasLocalMutations: true }
+              )
+            ],
+            hasPendingWrites: true
+          })
+          // Loading the bundle will not raise snapshots, because the
+          // mutation is not acknowledged.
+          .loadBundle(bundleString)
+      );
+    }
+  );
+
+  specTest('Newer docs from bundles might lead to limbo doc', [], () => {
+    const query = Query.atPath(path('collection'));
+    const docA = doc('collection/a', 1000, { key: 'a' });
+    const bundleString1 = bundleWithDocument({
+      key: docA.key,
+      readTime: 500,
+      createTime: 250,
+      updateTime: 500,
+      content: { key: { stringValue: 'b' } }
+    });
+
+    return (
+      spec()
+        .withGCEnabled(false)
+        .userListens(query)
+        .watchAcksFull(query, 250)
+        // Backend tells there is no such doc.
+        .expectEvents(query, {})
+        // Bundle tells otherwise, leads to limbo.
+        .loadBundle(bundleString1)
+        .expectEvents(query, {
+          added: [doc('collection/a', 500, { key: 'b' })],
+          fromCache: true
+        })
+        .expectLimboDocs(docA.key)
```

Comment:
Can you expand this test: We should verify that when the Limbo doc is resolved we go back to `fromCache: false`.

Reply:
Done.